File: src/lib/format.ts
import chalk from "chalk";
import Table from "cli-table3";
import type { Bookmark, Space, SessionMeta } from "../types.js";
import { shortSessionId } from "./sessionDisplay.js";

export function formatSessionTable(sessions: SessionMeta[]): string {
  const formatToken = (value: number | undefined): string => {
    return value === undefined ? "-" : String(value);
  };

  const table = new Table({
    head: [
      chalk.cyan("Session ID"),
      chalk.cyan("Agent"),
      chalk.cyan("Model"),
      chalk.cyan("Project"),
      chalk.cyan("Modified"),
      chalk.cyan("Input"),
      chalk.cyan("Output"),
      chalk.cyan("Total"),
      chalk.cyan("Cache"),
    ],
    colWidths: [15, 8, 16, 30, 20, 10, 10, 10, 10],
    style: { head: [] },
  });
  for (const s of sessions) {
    const shortId = shortSessionId(s.session_id);
    const agent = s.provider === "codex" ? "codex" : "claude";
    const shortProject = s.project_path ? (s.project_path.length > 36 ? "…" + s.project_path.slice(-35) : s.project_path) : "-";
    const shortDate = s.modified_at.slice(0, 19).replace("T", " ");
    table.push([
      shortId,
      agent,
      s.model || "-",
      shortProject,
      shortDate,
      formatToken(s.token_usage?.input_tokens),
      formatToken(s.token_usage?.output_tokens),
      formatToken(s.token_usage?.total_tokens),
      formatToken(s.token_usage?.cache_tokens),
    ]);
  }
  return table.toString();
}

export function formatBookmarkTable(bookmarks: Bookmark[]): string {
  const table = new Table({
    head: [chalk.green("ID"), chalk.green("Title"), chalk.green("Category"), chalk.green("Tags"), chalk.green("Created")],
    colWidths: [16, 30, 15, 20, 22],
    style: { head: [] },
  });
  for (const b of bookmarks) {
    const title = b.title.length > 28 ? b.title.slice(0, 28) + "…" : b.title;
    const tags = b.tags.join(", ") || "-";
    const date = b.created_at.slice(0, 19).replace("T", " ");
    table.push([b.id, title, b.category || "-", tags, date]);
  }
  return table.toString();
}

export function formatBookmarkDetail(b: Bookmark): string {
  const lines: string[] = [
    chalk.bold.green(`Pin: ${b.id}`),
    `  Title:       ${b.title}`,
    `  Provider:    ${b.provider}`,
    `  Session:     ${b.session_id}`,
    `  Category:    ${b.category || "(none)"}`,
    `  Tags:        ${b.tags.join(", ") || "(none)"}`,
    `  Project:     ${b.project_path}`,
    `  First Prompt:${b.first_prompt ? "\n" + wrapText(b.first_prompt, 60, "    ") : " (none)"}`,
    `  Created:     ${b.created_at}`,
    `  Updated:     ${b.updated_at}`,
  ];

  if (b.space_ids.length > 0) {
    lines.push(`  Catalogs:    ${b.space_ids.join(", ")}`);
  }

  if (b.notes.length > 0) {
    lines.push(chalk.bold("  Notes:"));
    for (const n of b.notes) {
      lines.push(`    [${n.id}] ${wrapText(n.content, 56, "      ")}`);
      lines.push(chalk.gray(`      ${n.created_at}`));
    }
  }

  return lines.join("\n");
}

export function formatSpaceTree(spaces: Space[], bookmarks: Bookmark[]): string {
  if (spaces.length === 0) return chalk.yellow("No catalogs created yet.");

  const childrenMap = new Map<string | null, Space[]>();
  for (const s of spaces) {
    const parent = s.parent_id ?? null;
    if (!childrenMap.has(parent)) childrenMap.set(parent, []);
    childrenMap.get(parent)!.push(s);
  }

  const bookmarkBySpace = new Map<string, Bookmark[]>();
  for (const b of bookmarks) {
    for (const sid of b.space_ids) {
      if (!bookmarkBySpace.has(sid)) bookmarkBySpace.set(sid, []);
      bookmarkBySpace.get(sid)!.push(b);
    }
  }

  function renderNode(space: Space, prefix: string, isLast: boolean): string[] {
    const connector = isLast ? "└── " : "├── ";
    const lines: string[] = [];
    const tagStr = space.tags.length > 0 ? chalk.gray(` [${space.tags.join(", ")}]`) : "";
    lines.push(`${prefix}${connector}${chalk.bold(space.name)}${tagStr}`);

    const childPrefix = prefix + (isLast ? "    " : "│   ");

    const bk = bookmarkBySpace.get(space.id) || [];
    for (let i = 0; i < bk.length; i++) {
      const bIsLast = i === bk.length - 1 && !childrenMap.has(space.id);
      const bConn = bIsLast ? "└── " : "├── ";
      lines.push(`${childPrefix}${bConn}${chalk.cyan(bk[i]!.title)} ${chalk.gray(`[${bk[i]!.session_id}]`)}`);
    }

    const children = childrenMap.get(space.id) || [];
    for (let i = 0; i < children.length; i++) {
      lines.push(...renderNode(children[i], childPrefix, i === children.length - 1 && bk.length === 0));
    }

    return lines;
  }

  const roots = childrenMap.get(null) || [];
  const lines: string[] = [chalk.bold("starling")];
  for (let i = 0; i < roots.length; i++) {
    lines.push(...renderNode(roots[i], "", i === roots.length - 1));
  }
  return lines.join("\n");
}

function wrapText(text: string, width: number, indent: string): string {
  const words = text.split(" ");
  const lines: string[] = [];
  let current = "";
  for (const word of words) {
    if ((current + " " + word).trim().length > width) {
      lines.push(current.trim());
      current = indent + word;
    } else {
      current += " " + word;
    }
  }
  if (current.trim()) lines.push(current.trim());
  return lines.join("\n");
}

File: rust/src/core/format.rs
//! Table rendering — mirrors src/lib/format.ts.

use std::collections::HashMap;

use colored::*;

use crate::core::runs::{status_badge, RunStatus};
use crate::core::session_display::short_session_id;
use crate::types::{Bookmark, SessionMeta, Space};

fn fmt_token(v: Option<u64>) -> String {
    match v {
        Some(n) => n.to_string(),
        None => "-".to_string(),
    }
}

fn truncate_left(s: &str, max: usize) -> String {
    let len = s.chars().count();
    if len <= max { return s.to_string(); }
    let suffix: String = s.chars().skip(len.saturating_sub(max - 1)).collect();
    format!("…{suffix}")
}

fn truncate_right(s: &str, max: usize) -> String {
    let len = s.chars().count();
    if len <= max { return s.to_string(); }
    let prefix: String = s.chars().take(max.saturating_sub(1)).collect();
    format!("{prefix}…")
}

/// Build the canonical session list table as plain text. `status_map` optional.
pub fn format_session_table(sessions: &[SessionMeta], status_map: Option<&HashMap<String, RunStatus>>) -> String {
    use comfy_table::{Cell, Color, ContentArrangement, Table, presets::UTF8_FULL};
    let has_status = status_map.is_some();
    let mut table = Table::new();
    table
        .load_preset(UTF8_FULL)
        .set_content_arrangement(ContentArrangement::Disabled);

    let mut headers: Vec<Cell> = vec![Cell::new("Session ID").fg(Color::Cyan)];
    if has_status { headers.push(Cell::new("Status").fg(Color::Cyan)); }
    headers.extend([
        Cell::new("Agent").fg(Color::Cyan),
        Cell::new("Model").fg(Color::Cyan),
        Cell::new("Project").fg(Color::Cyan),
        Cell::new("Modified").fg(Color::Cyan),
        Cell::new("Input").fg(Color::Cyan),
        Cell::new("Output").fg(Color::Cyan),
        Cell::new("Total").fg(Color::Cyan),
        Cell::new("Cache").fg(Color::Cyan),
    ]);
    table.set_header(headers);

    for s in sessions {
        let short_id = short_session_id(&s.session_id);
        let agent = if s.provider == "codex" { "codex" } else { "claude" };
        let short_project = if s.project_path.is_empty() {
            "-".to_string()
        } else {
            truncate_left(&s.project_path, 36)
        };
        let short_date = s.modified_at.chars().take(19).collect::<String>().replace('T', " ");
        let model = if s.model.is_empty() { "-".to_string() } else { s.model.clone() };

        let tokens = s.token_usage.as_ref();

        let mut row: Vec<Cell> = vec![Cell::new(short_id)];
        if has_status {
            let st = status_map.and_then(|m| m.get(&s.session_id)).copied().unwrap_or(RunStatus::Unknown);
            row.push(Cell::new(status_badge(st, false)));
        }
        row.extend([
            Cell::new(agent),
            Cell::new(model),
            Cell::new(short_project),
            Cell::new(short_date),
            Cell::new(fmt_token(tokens.and_then(|t| t.input_tokens))),
            Cell::new(fmt_token(tokens.and_then(|t| t.output_tokens))),
            Cell::new(fmt_token(tokens.and_then(|t| t.total_tokens))),
            Cell::new(fmt_token(tokens.and_then(|t| t.cache_tokens))),
        ]);
        table.add_row(row);
    }
    table.to_string()
}

pub fn format_bookmark_table(bookmarks: &[Bookmark]) -> String {
    use comfy_table::{Cell, Color, ContentArrangement, Table, presets::UTF8_FULL};
    let mut table = Table::new();
    table
        .load_preset(UTF8_FULL)
        .set_content_arrangement(ContentArrangement::Disabled);
    table.set_header(vec![
        Cell::new("ID").fg(Color::Green),
        Cell::new("Title").fg(Color::Green),
        Cell::new("Category").fg(Color::Green),
        Cell::new("Tags").fg(Color::Green),
        Cell::new("Created").fg(Color::Green),
    ]);

    for b in bookmarks {
        let title = truncate_right(&b.title, 28);
        let tags = if b.tags.is_empty() { "-".to_string() } else { b.tags.join(", ") };
        let category = if b.category.is_empty() { "-".to_string() } else { b.category.clone() };
        let date = b.created_at.chars().take(19).collect::<String>().replace('T', " ");
        table.add_row(vec![
            Cell::new(&b.id),
            Cell::new(title),
            Cell::new(category),
            Cell::new(tags),
            Cell::new(date),
        ]);
    }
    table.to_string()
}

pub fn wrap_text(text: &str, width: usize, indent: &str) -> String {
    let mut lines = Vec::new();
    let mut current = String::new();
    for word in text.split_whitespace() {
        let candidate = if current.is_empty() { word.to_string() } else { format!("{current} {word}") };
        if candidate.chars().count() > width {
            if !current.is_empty() { lines.push(std::mem::take(&mut current)); }
            current = format!("{indent}{word}");
        } else {
            current = candidate;
        }
    }
    if !current.trim().is_empty() { lines.push(current); }
    lines.join("\n")
}

pub fn format_bookmark_detail(b: &Bookmark) -> String {
    let mut lines: Vec<String> = vec![
        format!("{}", "Pin:".bold().green()),
        format!("  Title:       {}", b.title),
        format!("  Provider:    {}", b.provider),
        format!("  Session:     {}", b.session_id),
        format!("  Category:    {}", if b.category.is_empty() { "(none)".to_string() } else { b.category.clone() }),
        format!("  Tags:        {}", if b.tags.is_empty() { "(none)".to_string() } else { b.tags.join(", ") }),
        format!("  Project:     {}", b.project_path),
        format!("  First Prompt:{}",
            if b.first_prompt.is_empty() {
                " (none)".to_string()
            } else {
                format!("\n    {}", wrap_text(&b.first_prompt, 60, "    "))
            }),
        format!("  Created:     {}", b.created_at),
        format!("  Updated:     {}", b.updated_at),
    ];
    if !b.space_ids.is_empty() {
        lines.push(format!("  Catalogs:    {}", b.space_ids.join(", ")));
    }
    if !b.notes.is_empty() {
        lines.push(format!("{}", "  Notes:".bold()));
        for n in &b.notes {
            lines.push(format!("    [{}] {}", n.id, wrap_text(&n.content, 56, "      ")));
            lines.push(format!("      {}", n.created_at.normal()));
        }
    }
    lines.join("\n")
}

/// `starling` tree of catalogs + pinned sessions.
pub fn format_space_tree(spaces: &[Space], bookmarks: &[Bookmark]) -> String {
    if spaces.is_empty() {
        return "No catalogs created yet.".yellow().to_string();
    }

    let mut children_map: HashMap<Option<String>, Vec<&Space>> = HashMap::new();
    for s in spaces {
        children_map.entry(s.parent_id.clone()).or_default().push(s);
    }

    let mut bookmark_by_space: HashMap<String, Vec<&Bookmark>> = HashMap::new();
    for b in bookmarks {
        for sid in &b.space_ids {
            bookmark_by_space.entry(sid.clone()).or_default().push(b);
        }
    }

    // Stable sort children by name
    for v in children_map.values_mut() {
        v.sort_by(|a, b| a.name.cmp(&b.name));
    }

    fn render_node(
        space: &Space,
        prefix: &str,
        is_last: bool,
        children_map: &HashMap<Option<String>, Vec<&Space>>,
        bookmark_by_space: &HashMap<String, Vec<&Bookmark>>,
    ) -> Vec<String> {
        let connector = if is_last { "└── " } else { "├── " };
        let mut lines: Vec<String> = Vec::new();
        let tag_str = if space.tags.is_empty() {
            String::new()
        } else {
            format!(" {}", format!("[{}]", space.tags.join(", ")).normal())
        };
        lines.push(format!("{prefix}{connector}{}{tag_str}", space.name.bold()));

        let child_prefix = format!("{prefix}{}", if is_last { "    " } else { "│   " });

        let bk_opt = bookmark_by_space.get(&space.id);
        let bk_len = bk_opt.map(|v| v.len()).unwrap_or(0);
        let child_len = children_map.get(&Some(space.id.clone())).map(|v| v.len()).unwrap_or(0);

        if let Some(bk) = bk_opt {
            for (i, b) in bk.iter().enumerate() {
                let b_is_last = i + 1 == bk_len && child_len == 0;
                let b_conn = if b_is_last { "└── " } else { "├── " };
                lines.push(format!("{child_prefix}{b_conn}{} {}",
                    b.title.cyan(),
                    format!("[{}]", b.session_id).normal()));
            }
        }

        if let Some(children) = children_map.get(&Some(space.id.clone())) {
            for (i, child) in children.iter().enumerate() {
                let child_is_last = i + 1 == child_len && bk_len == 0;
                let nested = render_node(child, &child_prefix, child_is_last, children_map, bookmark_by_space);
                lines.extend(nested);
            }
        }

        lines
    }

    let mut lines: Vec<String> = vec!["starling".bold().to_string()];
    let roots = children_map.get(&None).cloned().unwrap_or_default();
    for (i, root) in roots.iter().enumerate() {
        let is_last = i + 1 == roots.len();
        lines.extend(render_node(root, "", is_last, &children_map, &bookmark_by_space));
    }
    lines.join("\n")
}
